vm/src/obj/objint.rs

```diff
@@ -810,79 +786,160 @@ pub fn to_int(vm: &VirtualMachine, obj: &PyObjectRef, base: &BigInt) -> PyResult
-                    buf.remove(0);
+fn to_int_radix(vm: &VirtualMachine, obj: &PyObjectRef, base: u32) -> PyResult<BigInt> {
+    debug_assert!(base == 0 || (2..=36).contains(&base));
+
+    let opt = match_class!(match obj.clone() {
+        string @ PyString => {
+            let s = string.as_str();
+            bytes_to_int(s.as_bytes(), base)
+        }
+        bytes @ PyBytes => {
+            let bytes = bytes.get_value();
+            bytes_to_int(bytes, base)
+        }
+        bytearray @ PyByteArray => {
+            let inner = bytearray.borrow_value();
+            bytes_to_int(&inner.elements, base)
+        }
+        _ => {
+            return Err(
+                vm.new_type_error("int() can't convert non-string with explicit base".to_owned())
+            );
+        }
+    });
+    match opt {
+        Some(int) => Ok(int),
+        None => Err(vm.new_value_error(format!(
+            "invalid literal for int() with base {}: {}",
+            base,
+            vm.to_repr(obj)?,
+        ))),
+    }
+}
+
+fn bytes_to_int(lit: &[u8], mut base: u32) -> Option<BigInt> {
+    // split sign
+    let mut lit = lit.trim();
+    let sign = match lit.first()? {
+        b'+' => Some(Sign::Plus),
+        b'-' => Some(Sign::Minus),
+        _ => None,
+    };
+    if sign.is_some() {
+        lit = unsafe {
+            // safe by sign check
+            lit.get_unchecked(1..)
+        };
+    }
+
+    // split radix
+    let first = *lit.first()?;
+    let has_radix = if first == b'0' {
+        match base {
+            0 => {
+                if let Some(parsed) = lit.get(1).and_then(detect_base) {
+                    base = parsed;
+                    true
+                } else {
+                    if let Some((last, body)) = lit[1..].split_last() {
+                        let is_zero = body.iter().all(|&c| c == b'0' || c == b'_') && *last == b'0';
+                        if !is_zero {
+                            return None;
+                        }
+                    }
+                    return Some(BigInt::zero());
                 }
-            } else if (matched_radix == 2 && base < 12)
-                || (matched_radix == 8 && base < 25)
-                || (matched_radix == 16 && base < 34)
-            {
-                return None;
             }
+            16 => lit.get(1).map_or(false, |&b| matches!(b, b'x' | b'X')),
+            2 => lit.get(1).map_or(false, |&b| matches!(b, b'b' | b'B')),
+            8 => lit.get(1).map_or(false, |&b| matches!(b, b'o' | b'O')),
+            _ => false,
+        }
+    } else {
+        if base == 0 {
+            base = 10;
+        }
+        false
+    };
+    if has_radix {
+        lit = unsafe {
+            // safe by radix check
+            lit.get_unchecked(2..)
+        };
+        if lit.first() == Some(&b'_') {
+            lit = unsafe {
+                // safe by b'_' check
+                lit.get_unchecked(1..)
+            };
```

Comment:
```suggestion
        lit = &lit[2..];
        if lit.first()? == &b'_' {
            lit = &lit[1..];
```
This has no bounds checks; if there's nothing after `0x` it should be an error anyway

Reply:
nIce catch!